File: android/app/src/main/java/com/pscheol/ionicsample/plugins/SocialLoginPlugin.kt
package com.pscheol.ionicsample.plugins

import android.Manifest
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.PluginResult
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.kakao.sdk.auth.model.OAuthToken
import com.kakao.sdk.user.UserApiClient
import java.util.*


@CapacitorPlugin(
    name = "SocialLogin"
)
class SocialLoginPlugin : Plugin() {

    @PluginMethod
    fun isKakaoTalkLoginAvailable(call: PluginCall) {
        val ret = UserApiClient.instance.isKakaoTalkLoginAvailable(activity)
        val obj = JSObject()
        obj.put("available", ret)
        call.resolve(obj)
    }

    private fun loginResult(call: PluginCall, token: OAuthToken?, error: Throwable?) {
        if(error != null) {
            call.reject(error.message)
        }
        else if(token == null) {
            call.reject("")
        }
        else {
            val obj = JSObject()
            obj.put("accessToken", token.accessToken)
            obj.put("accessTokenExpiresAt", token.accessTokenExpiresAt.toString())
            obj.put("refreshToken", token.refreshToken)
            obj.put("refreshTokenExpiresAt", token.refreshTokenExpiresAt.toString())
            obj.put("idToken", token.idToken)
            obj.put("scopes", token.accessToken)

            call.resolve(obj)
        }
    }

    @PluginMethod
    fun loginWithKakaoTalk(call: PluginCall) {
        UserApiClient.instance.loginWithKakaoTalk(activity) { token, error ->
            loginResult(call, token, error)
        }
    }

    @PluginMethod
    fun loginWithKakaoAccount(call: PluginCall) {
        UserApiClient.instance.loginWithKakaoAccount(activity) { token, error ->
            loginResult(call, token, error)
        }
    }
}

File: ios/App/App/Plugins/SocialLoginPlugin.swift
//
//  SocialLogin.swift
//  App
//
//  Created by 모비핀 on 2022/08/30.
//

import Capacitor
import KakaoSDKUser
import KakaoSDKAuth

@objc(SocialLoginPlugin)
public class SocialLoginPlugin : CAPPlugin {
    
    @objc public func isKakaoTalkLoginAvailable(_ call: CAPPluginCall) {
        let ret = KakaoSDKUser.UserApi.isKakaoTalkLoginAvailable()
        
        call.resolve(["available" : ret])
    }
    
    private func loginResult(_ call: CAPPluginCall, token: OAuthToken?, error: Error?) {
        
        
        if let error = error {
            print(error)
            call.reject("")
            
        }
        else if token == nil {
            call.reject("")
        }
        else {
            print("loginWithKakaoTalk() success.")

            let token: OAuthToken = token!
            
            let obj = [
                "accessToken" : token.accessToken,
                "accessTokenExpiresAt" : token.expiredAt.toString(),
                "refreshToken" : token.refreshToken,
                "refreshTokenExpiresAt" : token.refreshTokenExpiredAt.toString(),
//                "idToken", token.idToken,
                "scopes" : token.accessToken
            ]
            
            
            call.resolve(obj)
        }
    }
    
    
    @objc public func loginWithKakaoTalk(_ call: CAPPluginCall) {
        
        if(!UserApi.isKakaoTalkLoginAvailable()) {
            call.reject("")
            return
        }
        
                
        UserApi.shared.loginWithKakaoTalk {(oauthToken, error) in
            
            self.loginResult(call, token: oauthToken, error: error)
        }
    }
    
    @objc public func loginWithKakaoAccount(_ call: CAPPluginCall) {
        UserApi.shared.loginWithKakaoAccount {(oauthToken, error) in
            
            self.loginResult(call, token: oauthToken, error: error)
        }
    }
}


